app/deliveryoptions/utils.py

```diff
@@ -0,0 +1,727 @@
+
+
+def html_replacer(string: str, record: dict, surrogate_data: List) -> str:
+    subs = re.findall(r"{[A-Za-z]*}", string)
+
+    for s in subs:
+        try:
+            func = deliveryOptionsTags[s]
+
+            # If the tag doesn't have any data (can happen with surrogate links),
+            # rather than have a string with something missing, just return an
+            # empty string and it won't get displayed
+            if f := func(record, surrogate_data):
+                string = string.replace(s, f)
+        except Exception:
+            raise
+    return string
+
+
+def html_builder(
+    delivery_option_data: Union[List, str],
+    record_data: Dict,
+    surrogate_data: List = [],
+    dcs: bool = False,
+) -> str:
+    html = ""
+
+    if delivery_option_data is None:
+        return html
+
+    if isinstance(delivery_option_data, list):
+        for item in delivery_option_data:
+            if not dcs and item["name"] == "descriptionDCS":
+                pass
+            else:
+                html += html_replacer(
+                    item["value"], record_data, surrogate_data
+                )
+    else:
+        html = html_replacer(delivery_option_data, record_data, surrogate_data)
+
+    return html
+
+
+# Specific pre-processing for the order buttons data
+def orderbuttons_builder(
+    delivery_option_data: List, record_data: Dict, surrogate_data: List
+) -> List:
+    for item in delivery_option_data:
+        item["href"] = html_builder(
+            item["href"], record_data, surrogate_data=surrogate_data
+        )
+        item["text"] = html_builder(
+            item["text"], record_data, surrogate_data=surrogate_data
+        )
+    return delivery_option_data
+
+
+# Specific pre-processing for the basket limit data
+def basketlimit_builder(
+    delivery_option_data: Union[List, str], record_data: Dict
+) -> str:
+    return html_builder(delivery_option_data, record_data)
+
+
+# Specific pre-processing for the expand link data
+def expandlink_builder(
+    delivery_option_data: Union[List, str], record_data: Dict
+) -> str:
+    return html_builder(delivery_option_data, record_data)
+
+
+# Specific pre-processing for the description data
+def description_builder(
+    delivery_option_data: Union[List, str],
+    record_data: Dict,
+    surrogate_data: List,
+) -> str:
+    if distressing_content_match(record_data.reference_number):
+        return html_builder(
+            delivery_option_data,
+            record_data,
+            surrogate_data=surrogate_data,
+            dcs=True,
+        )
+
+    return html_builder(
+        delivery_option_data, record_data, surrogate_data=surrogate_data
+    )
+
+
+# Specific pre-processing for the supplemental data
+def supplemental_builder(
+    delivery_option_data: Union[List, str],
+    record_data: Dict,
+    surrogate_data: List,
+) -> str:
+    return html_builder(
+        delivery_option_data, record_data, surrogate_data=surrogate_data
+    )
```

Comment:
see comments for functions `expandlink_builder, basketlimit_builder`

Reply:
Relocated outdated comment https://github.com/nationalarchives/ds-search/pull/25/files#r2012458471